NeoML/include/NeoML/Dnn/DnnSimpleAPI.h

```diff
@@ -0,0 +1,512 @@
+	const CLayerOutput& layer5, const CLayerOutput& layer6 )
+{
+	NeoAssert( !layer1.IsOptional() );
+	NeoAssert( layer1.IsValid() );
+	NeoAssert( name != 0 );
+
+	if( layer == 0 ) {
+		CDnn* network = layer1.Layer->GetDnn();
+		layer = new T( network->GetMathEngine() );
+		if( !initFunc.IsEmpty() ) {
+			initFunc( layer );
+		}
+		layer->SetName( name );
+		network->AddLayer( *layer );
+	}
+
+	CArray<CLayerOutput> inputLayers;
+	inputLayers.Add( layer1 );
+	inputLayers.Add( layer2 );
+	inputLayers.Add( layer3 );
+	inputLayers.Add( layer4 );
+	inputLayers.Add( layer5 );
+	inputLayers.Add( layer6 );
+
+	const int startIndex = layer->GetInputCount();
+	for( int i = 0; i < inputLayers.Size(); i++ ) {
+		const CLayerOutput& inputLayer = inputLayers[i];
+		if( inputLayer.IsOptional() ) {
+			break;
+		}
+		NeoAssert( inputLayer.IsValid() );
+		layer->Connect( startIndex + i, *inputLayer.Layer, inputLayer.OutputNumber );
+	}
+
+	return layer;
+}
+
+template<typename T>
+T* CLayerWrapper<T>::operator()(
+	const CLayerOutput& layer1, const CLayerOutput& layer2,
+	const CLayerOutput& layer3, const CLayerOutput& layer4,
+	const CLayerOutput& layer5, const CLayerOutput& layer6 )
+{
+	NeoAssert( layer1.IsValid() );
+	CDnn* network = layer1.Layer->GetDnn();
+	const CString name = findFreeLayerName( *network, prefix );
+	return operator()( name, layer1, layer2, layer3, layer4, layer5, layer6 );
+}
+
+template<typename T>
+CString CLayerWrapper<T>::findFreeLayerName(
+	const CDnn& network, const char* prefix ) const
+{
+	const CString prefixStr( prefix );
+
+	int index = 0;
+	while( true ) {
+		const CString newName = prefixStr + "_" + Str( index++ );
+		if( !network.HasLayer( newName ) ) {
+			return newName;
+		}
+	}
+	NeoAssert( false );
+	// make compiler happy
+	return CString();
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+
+/*
+Simple API for create neural networks.
+
+Example 0:
+
+	CDnn net{ rnd, engine };
+
+	auto x = Source( net, "Input0" );
+	auto labels = Source( net, "Labels" );
+
+	x = FullyConnected( 100 )( x );
+	x = Relu() ( x );
+	x = FullyConnected( 200 )( x );
+	x = Gelu()( x );
+	x = Dropout( 0.5f )( x )
+	x = FullyConnected( 1 )( x );
+	BinaryCrossEntropy()( x, labels );
+
+Example 2:
+
+	CDnn net{ rnd, engine };
+
+	auto x = Source( net, "Input0" );
+	auto y = Source( net, "Input1" );
+	auto labels = Source( net, "Labels" );
+	auto weights = Source( net, "Weightss" );
+
+	auto fc = FullyConnected( 100 );
+
+	x = fc( x ); // 1.
+	y = fc( y ); // 2. Share weights with 1.
+	x = Concat() ( x, y );
+	x = Relu() ( x );
+	x = fc( x );
+	x = Gelu()( x );
+	x = Dropout( 0.5f )( x )
+	x = FullyConnected( 1 )( x );
+	BinaryCrossEntropy( 2.0f )( x, labels, weights );
+
+*/
+
+
+//////////////////////////////////////////////////////////////////////////////////////////
+
+// Create CSourceLayer with name
+NEOML_API CBaseLayer* Source( CDnn& network, const char* name );

```

Comment:
This is for convenience. 

For example:

**with CBaseLayer:**
auto x = Source( network, "Input" );
x = FullyConnected( 1024 ) (x);
x = Relu()(x);

**with CSourceLayer:**
auto x = Source( network, "Input" );
x = FullyConnected( 1024 ) (x); // Error cannot convert from 'NeoML::CFullyConnectedLayer *' to 'NeoML::CSourceLayer *'
x = Relu()(x);

Reply:
But, as I said earlier, how do you gonna call `CSourceLayer::SetBlob` afterwards?

In your examples you can simply replace `auto` with `CBaseLayer*` and it will be working in both cases (`CBaseLayer* Source(...)` and `CSourceLayer* Source(...)`).

Reply:
I don't want to mix create network function with blob filling. So we don't need to use the input layer as soon as we created it.

CDnn network{...};
buildNetwordVersion1( network );
trainNetworkEpoch(network);
testNetwork(network);